runtime/drivers/clickhouse/olap.go

```diff
@@ -251,7 +251,94 @@ func (c *connection) InsertTableAsSelect(ctx context.Context, name, sql string,
 		})
 	}
 
-	// merge strategy is also not supported for clickhouse
+	if strategy == drivers.IncrementalStrategyMerge {
+		_, onCluster, err := informationSchema{c: c}.entityType(ctx, "", name)
+		if err != nil {
+			return err
+		}
+		onClusterClause := ""
+		if onCluster {
+			onClusterClause = "ON CLUSTER " + safeSQLName(c.config.Cluster)
+		}
+		// Get the engine info of the given table
+		var engine string
+		res, err := c.Execute(ctx, &drivers.Statement{
+			Query:    "SELECT engine FROM system.tables WHERE database = currentDatabase() AND name = ?",
+			Args:     []any{name},
+			Priority: 1,
+		})
+		if err != nil {
+			return err
+		}
+		defer res.Close()
+		if res.Next() {
+			if err := res.Scan(&engine); err != nil {
+				return err
+			}
+		}
+		// Distributed table cannot be altered directly, so we need to alter the local table
+		if engine == "Distributed" {
+			name = localTableName(name)
+		}
+		// create temp table with the same schema
+		tempName := tempName(name)
+		err = c.Exec(ctx, &drivers.Statement{
+			Query:    fmt.Sprintf("CREATE OR REPLACE TABLE %s %s AS %s", safeSQLName(tempName), onClusterClause, name),
+			Priority: 1,
+		})
+		if err != nil {
+			return err
+		}
+		// clean up the temp table
+		defer func() {
+			err = c.Exec(ctx, &drivers.Statement{
+				Query:    fmt.Sprintf("DROP TABLE %s %s", safeSQLName(tempName), onClusterClause),
+				Priority: 1,
+			})
+			if err != nil {
+				c.logger.Warn("clickhouse: failed to drop temp table", zap.String("name", tempName), zap.Error(err))
+			}
+		}()
+		// insert into temp table
+		err = c.Exec(ctx, &drivers.Statement{
+			Query:    fmt.Sprintf("INSERT INTO %s %s", safeSQLName(tempName), sql),
+			Priority: 1,
+		})
+		if err != nil {
+			return err
+		}
+		// list partitions from the temp table using system.parts
+		res, err = c.Execute(ctx, &drivers.Statement{
+			Query:    fmt.Sprintf("SELECT DISTINCT partition FROM system.parts WHERE table = '%s'", tempName),
+			Priority: 1,
+		})
+		if err != nil {
+			return err
+		}
+		defer res.Close()
```

Comment:
This still keeps the connection blocked till this function returns. You can use the same approach as noted above.
May be I am overthinking and it is fine to keep this connection blocked since the subsequent queries are going to be fast.